Adi/document (#59)
adipill04/ge-healthcare: GEHC-PreInstall-Starter-Code/README.md

# GE Health - Pre-installation Project: Team 2

## Project requirements
### Task tracker:
https://docs.google.com/spreadsheets/d/1gf_UTF-r0HorI7Kt7kTUibhkB9YY_gbjp9QXbJiUSPU/edit?usp=sharing

### User story document:

> https://uwmadison.app.box.com/file/1650319754225?s=6w7eluows0osywq3hms8pu53kjcbtpxf

Reviewer: ```suggestion
* [Task tracker](https://docs.google.com/spreadsheets/d/1gf_UTF-r0HorI7Kt7kTUibhkB9YY_gbjp9QXbJiUSPU/edit?usp=sharing)
* [User story document](https://uwmadison.app.box.com/file/1650319754225?s=6w7eluows0osywq3hms8pu53kjcbtpxf)
```

 > 1. Make sure the python3 path is correct. You can verify this by running `which python3` in the CLI.
 > 2. Make sure the app.py path is correct.

## User Requirements

> Please refer to [user requirements](https://uwmadison.app.box.com/file/1650319754225?s=6w7eluows0osywq3hms8pu53kjcbtpxf).

Reviewer: May I know why we're planning on removing this section? Thanks!

Author: Private doc, don't think it's necessary for a readme

Reviewer: It asks for GE-healthcare credentials, which is fine considering how this is supposed to be an internal tool

![image](https://github.com/user-attachments/assets/b2b9f6d8-32ef-4e32-b29b-3bbc91f41dde)

Author: Ok, sounds good to me then!

Reviewer: Great, marking this as resolved.

--> At most will take 5 seconds to run (time it takes the latency to **time out**).

### Firewall and Security Application
--> Tests reachability of entered IP addresses
--> Able to read files from Firewall_Preconfigs directory 

## Known bugs:
1. Thread destroyed error<br>
   --> Very rare error. Occurs randomly when there is a conflict between multiple workers or the main thread.<br>
   --> Fix: Make app more thread safe by adding locks in non-graphing functions<br>
2. Pausing, then stopping a live test (RSSI or Performance) causes app to crash<br>
   --> Fix: add graceful termination logic for when the app is paused<br>
3. DHCP table not populating with data<br>
   --> The parser fails to return parsed data. DHCP test works as a whole (get data from tcpdump -> parse data -> display data), however it's possible due to differences in stdout from tcpdump when running the test in a different place causes the parser (which uses regex) to fail? This bug was found last minute and not enough testing was done to confirm this.<br>
   --> Fix: start with reproducing the bug. Test the regex for the respective tcpdump stdout.<br>

> 4. Pulling system logs for a given time doesn't work correctly. Most of the time logs don't show up at all. Not much testing was done regarding the bug, so further investigation is required.

Reviewer: ```suggestion
4. Pulling system logs for a given time doesn't work correctly. Most of the time logs don't show up at all. Limited testing was done regarding the bug, so further investigation is required.
```

--> Currently, this process uses a QThread worker. Since the app could do a little better in terms of being thread safe, we recommend using multiprocessing instead for this. Using multiprocessing for longer running background programs makes life much easier!<br>

### 2. Automating connecting to a network 
--> Instead of the user having to connect each time the app starts, automating this using a config file might be more efficient

### 3. Updating version of Python without GIL
--> At the time of writing this, Python has released a no GIL experimental mode as part of Python 3.13. A stable release for this mode might be more desired.<br>
--> Having true multithreading will make the app a lot faster, especially when wanting to have a graph for each AP.<br>

### 4. App becomes slightly laggy when reaching the value for time window, entered by the user (General & Performance tests)
--> Although the graphing process is completely done in a thread, there is a small lag up until the graph reaches the value entered in the time window. This means the smaller the time window value, the less time the app takes to get back to its normal speed.<br>

> --> Recomendation for fixing this would be changing the logic behind the graph time window, or as stated above in _3_, true multithreading would probably fix it. <br>

Reviewer: ```suggestion
--> Recommendation for fixing this would be changing the logic behind the graph time window, or as stated above in _3_, true multithreading would probably fix it. <br>
```

## Features
### General Network Performance

| **Description**                  | **Status**       | **Notes**                                         |
|----------------------------------|------------------|--------------------------------------------------|
| As a user, I want to be able to see a graph of latency over time for a given test run.     | Done ||
| As a user, I want to be able to see a graph of packet loss over time for a given test run.              | Done            | |
| As a user, I want to be able to determine the number of hops between endpoints for a given test run.               | In Progress      |                   |
| As a user, I want to determine the MTU setting via MTU path discovery across a WAN connection   | Backlog            |   |

---
### Wi-Fi Performance

> | **Description**                  | **Status**       | **Notes**                                         |

Reviewer: @adipill04 We could either get rid of the entire section or keep them in here and mark their status completed. I'm leaning more towards the latter, as it doesn't require the user to switch back-and-forth between the documentation and the requirements, but I'm open to suggestions.

Author: For a readme I don't feel like we need to add documentation about functions or specific user stories. It's too much clutter. 

If we do want to add additional documentation I prefer it to be in the code

Reviewer: You're right that having both functions and user stories would make it cluttered. How about we have user stories in the README and function-related documentation within the functions?
